### ISSUE TITLE
Mark more characters as safe when passing build options

### ISSUE THREAD

**Issue by SemMulder** [inducer/pyopencl#250]
Right now the characters which are marked as safe [here](https://github.com/inducer/pyopencl/blob/4c0521913881ac6e284b38c9aab79708dfb522db/pyopencl/__init__.py#L392), are not exhaustive. Since we aren't actually passing the build options through a shell, but are invoking functions like `clBuildProgram`, the characters `-_{}()` are also safe.

I ran into this when compilation failed when trying to pass an array literal using `-D ARRAY_LITERAL={0,1}`. This just works if we mark the mentioned characters as safe.

I have a fix living [here](https://github.com/SemMulder/pyopencl), for which I can't open a PR on https://gitlab.tiker.net/inducer/pyopencl since I don't have an account.

Thoughts?

**Comment by inducer** [comment]
Thanks for the report and the suggested fix.

What'll happen if "unsafe" characters are found is that the affected argument gets quoted. Are you saying that the quoting is breaking not working in your use case.

**Comment by SemMulder** [comment]
Precisely! So the quoting results in the following:
```
<built-in>:9:9: error: macro name must be an identifier
In file included from stringInput.cl:2:
main.cl:193:57: error: use of undeclared identifier 'NEIGHBOURHOOD_ARRAY'
main.cl:348:57: error: use of undeclared identifier 'NEIGHBOURHOOD_ARRAY'

(options: ... -D 'NEIGHBOURHOOD_ARRAY={{-1,0},{0,-1},{0,1},{1,0}}' ...)
```

Note the quotes in the output. This happens with every OpenCL runtime I tested, so I don't think it's a bug in the runtime.

I checked the usages of [`_process_build_options`](https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L417) (which is the only usage of [`_shlex_qoute`](https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L395). They are
- https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L475,
- https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L583,
- https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L559.

None of them need _shell_ escaping, as far as I can tell. Since they all call `clBuildProgram` and `clLinkProgram` directly (i.e. not via a shell).

The quoting was introduced in commit 9b02b38887aa9a6851bcf34853ff8a07bbc638d5 which leads to PR https://github.com/inducer/pyopencl/pull/120. So I think having only the space as unsafe character would be sufficient. Because if you think about how the OpenCL runtime will parse the build options bytes, space will be used as argument separator and the rest it will just copy as is. In particular we don't need to escape things like `{`, `(` or `[` since we don't pass through a shell. So they don't have any meaning anywhere on the way from build option bytes to the actual parser.

I can prepare a PR for only marking the space as unsafe if you want? It feels kind of dangerous to just change the unsafe characters for all users, but I'm confident this makes way more sense :).

**Comment by inducer** [comment]
Yeah, I can confirm your observation. So it seems the idea of quoting is just thoroughly broken. Which begs the question what the correct behavior would be if PyOpenCL encounters a space in one of the options passed. Quoting is unlikely to help. Backslash escape?

**Comment by SemMulder** [comment]
Looking at [the ROCm OpenCL build options parser](https://github.com/RadeonOpenCompute/ROCm-OpenCL-Runtime/blob/master/compiler/lib/utils/options.cpp#L607) it indeed looks like the only unsafe character is a space, and it should be escaped by double quotes. I couldn't find any references to backslash escaping, so I'm pretty sure that wouldn't work.

The question now is how to implement this. Personally I like delegating the escaping to the user of `_process_build_options`. Which would mean within `pyopencl` making sure all arguments [here](https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L437), are quoted properly. Which comes down to quoting [`_DEFAULT_INCLUDE_OPTIONS`](https://github.com/inducer/pyopencl/blob/master/pyopencl/__init__.py#L266), if there is a space in there.

Delegating the quoting to the user might be a breaking change of course. Doing this automatically is a little bit error prone, since we can't distinguish between values and options. However, we might opt to just quote any string with a space in it. But passing a space if we really want to pass a space then becomes impossible.

What do you think?

**Comment by inducer** [comment]
The core issue is that PyOpenCL needs to be able to combine sets of build options (its own default ones and the user-supplied ones, for instance). Since the current quoting scheme doesn't seem to work, we're not really breaking anything, though it is worth looking back at the change that introduced the quoting to see what concrete issue (if any) provided the motivation. (It looks like #122 was that motivation, which provides further evidence that "simply glue existing build options together with spaces and rely on the user to quote" may be the better approach)

Since the current quoting scheme doesn't seem to work, I'd be willing to take the stance that we're not breaking much, and work towards a sane scheme instead.

In short, I think I'm mostly open to what you're suggesting here. I'd welcome a PR.

**Comment by SemMulder** [comment]
I finally managed to squeeze out a PR, please have a look.
